subsys/settings: Reworked settings module

Laczen opened the pull request:
Fixes#12160 - settings NVS backend

To allow easier integration of backends the settings module has been
reworked. Major changes are:

1. Change to the api: the set() routines are now of the form:
	''set(int argc, char **argv, size_t len, read_cb read, void *store)''
	the length that was found in the backend in len, and the data
	can be read using: ''read(store, data, len);
2. The reading of data in the set routines is directly reading in the
backend.
3. The records no longer need to be in the line format (the backend can
decide the best way to store them)
4. The runtime interface has been changed into a backend.
5. The settings module supports a default backend, this can be of type
fcb, file or nvs.
6. The user can setup any combination of backends when not using a
default backend (for this a settings_backend_init() routine has to be
provided).

Laczen committed:
subsys/settings: Reworked settings module

Modification to bluetooth settings related files.

Added bluetooth/host/mesh/settings.c

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

Laczen committed:
subsys/settings: Add some samples

Some samples are added to show how to use the settings module.

Corrections applied after review, added settings.rst at the
subsys/samples/settings level

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

dbkinder commented on samples/subsys/settings/custom/README.rst:
```diff
@@ -0,0 +1,32 @@
+This sample can be found under :file:`samples/subsys/settings/custom` in the
+Zephyr tree.
+
+The sample can be build for several platforms, the following commands build the
```
```suggestion
The sample can be built for several platforms. The following commands build the
```

Laczen replied:
Changed

dbkinder commented on samples/subsys/settings/custom/README.rst:
```diff
@@ -0,0 +1,32 @@
+
+The sample can be build for several platforms, the following commands build the
+application for the nrf51_pca10028 board. For the nrf51 the selected backend can
+only be fcb or nvs.
```
```suggestion
only be FCB or NVS.
```

Laczen replied:
changed

dbkinder commented on samples/subsys/settings/default/README.rst:
```diff
@@ -0,0 +1,31 @@
+
+This sample can be found under :file:`samples/subsys/settings` in the Zephyr tree.
+
+The sample can be build for several platforms, the following commands build the
```
```suggestion
The sample can be built for several platforms. The following commands build the
```

Laczen replied:
changed

dbkinder commented on samples/subsys/settings/default/README.rst:
```diff
@@ -0,0 +1,31 @@
+
+The sample can be build for several platforms, the following commands build the
+application for the nrf51_pca10028 board. For the nrf51 the selected backend can
+only be fcb or nvs.
```
```suggestion
only be FCB or NVS.
```

Laczen replied:
changed

dbkinder commented on samples/subsys/settings/default/README.rst:
```diff
@@ -0,0 +1,31 @@
+.. _settings-sample:
```
These document labels must be unique across all documents (the sample documentation above has the same label).  You'll need to change one of them.  That said, the words in these documents are identical (except for the document title).  What's up?

Laczen replied:
Hi @dbkinder, thanks for the review. I have changed the labels to settings-sample-default and settings-sample-custom. The difference between the two samples is that the default sample uses a default backend (selected by the config file), the second sample does a setup of a non-default backend that can be more suited to the needs.

dbkinder reviewed (approved):
OK... might be worth adding cross-references between these two sample docs.

Laczen committed:
subsys/settings: Reworked after review

Changes after review applied

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

fnde-ot commented on subsys/settings/settings_store.c:
```diff
@@ -0,0 +1,115 @@
+		prev = cur;
+	}
+
+	sys_slist_insert(&settings_load_srcs, prev, &cs->cs_next);
```
maybe you can just use sys_slist_append

fnde-ot commented on subsys/settings/settings_init.c:
```diff
@@ -0,0 +1,41 @@
+	static bool settings_initialized;
+	int rc;
+
+	if (settings_initialized) {
```
this check makes the "settings_cmd_inited" check in settings_init() not needed anymore

fnde-ot commented on subsys/settings/settings_store.c:
```diff
@@ -0,0 +1,115 @@
+	if (!cs) {
+		return -ENOENT;
+	}
+	return cs->cs_itf->csi_save(cs, name, (char *)value, val_len);
```
a couple of checks could be good before saving:
- strlen(name) against SETTINGS_MAX_NAME_LEN
- val_len against SETTINGS_MAX_VAL_LEN

not sure if similar checks should be done on the backends after loading...

fnde-ot reviewed (approved):
Great improvements to the settings subsystem, my comments are mostly nitpicking and a few minor suggestions.

fnde-ot commented:
It also seems that subsys/bluetooth/host/keys.c as well as a few tests use the "settings_val_get_len_cb" function that does not exist anymore!

Laczen commented:
@fnde-ot, thanks for the review. I will have a look at your comments and correct your proposals.

nvlsianpu commented:
I believe that I already revived the firs commit I reviewed https://github.com/zephyrproject-rtos/zephyr/pull/12892

Laczen commented:
@nvlsianpu, yes you did, remarks were taken into consideration. Except the intermediate line representation that I found difficult to understand/debug and would probably be difficult if both fcb and nffs are selected at the same time.

Laczen commented on include/settings/settings.h:
```diff
@@ -193,33 +181,6 @@ int settings_get_value(char *name, char *buf, int buf_len);
  */
 int settings_commit(char *name);
 
```
No, this is not possible because it calls settings_parse and lookup that changes name, could be done with a local name but I prefer to keep as it was.

fnde-ot replied:
You are right!

Laczen commented on include/settings/settings.h:
```diff
@@ -233,6 +194,79 @@ struct settings_store {
+int settings_runtime_get(const char *name, void *data, size_t len);
+
+#endif /* CONFIG_SETTINGS_RUNTIME */
+
```
I would rather change the settings_commit(name) to runtime_commit(name) and remove the settings_get() so that all is in the runtime interface. OK?

fnde-ot replied:
Sounds good to me, just watch out that settings_store.c is using settings_commit (line 114), so it probably needs to be enabled by default.

Laczen commented on samples/subsys/settings/custom/README.rst:
```diff
@@ -0,0 +1,32 @@
+********
+
+This is a simple application demonstrating use of the settings
+module for non-volatile (flash) storage.  The application demonstrates the
```
changed

Laczen commented on samples/subsys/settings/default/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+cmake_minimum_required(VERSION 3.8.2)
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
+project(nsetting)
```
changed

Laczen commented on samples/subsys/settings/default/README.rst:
```diff
@@ -0,0 +1,31 @@
+********
+
+This is a simple application demonstrating use of the settings
+module for non-volatile (flash) storage.  The application demonstrates the
```
changed

Laczen commented on samples/subsys/settings/default/prj.conf:
```diff
@@ -0,0 +1,41 @@
+#CONFIG_NFFS_FILESYSTEM_MAX_AREAS=8
+
+CONFIG_SETTINGS=y
+#CONFIG_SETTINGS_FCB=y
```
Selecting a default backend also selects the specified backend

fnde-ot replied:
Ok.

fnde-ot commented on subsys/settings/settings_priv.h:
```diff
@@ -0,0 +1,46 @@
+};
+
+void settings_src_register(struct settings_store *cs);
+void settings_dst_register(struct settings_store *cs);
```
It would also be nice to expose "settings_parse_and_lookup" for the backend to be able to find setting handlers when loading.

Laczen replied:
@fnde-ot, do you really need access to the variable sys_slist_t settings_handlers ?

fnde-ot replied:
Well I would need at least a way to locate a handler given a name, so if you prefer not exposing the sys_slist_t settings_handlers, I would need the settings_parse_and_lookup().

Laczen replied:
Thanks for the quick answer, settings_parse_and_lookup() is exposed.

Laczen commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2702,9 +2703,8 @@ static int ccc_set(int argc, char **argv, void *val_ctx)
-		len = settings_val_read_cb(val_ctx, ccc_store,
-					   sizeof(ccc_store));
+	if (len_rd) {
+		len = read(store, ccc_store, sizeof(ccc_store));
```
@fnde-ot, during the rework of all bluetooth settings related items I have taken the approach:
if there was a check on the length I changed it, if there was no check I didn't add the check.

fnde-ot replied:
Agreed, those improvements probably belong in another PR.

Laczen commented on subsys/bluetooth/host/settings.c:
```diff
@@ -114,7 +115,7 @@ static int set(int argc, char **argv, void *value_ctx)
 		len = sizeof(bt_dev.id_addr);
 
-		len = settings_val_read_cb(value_ctx, &bt_dev.id_addr, len);
+		len = read(store, &bt_dev.id_addr, len_rd);
```
There was no check, so I didn't add it

Laczen commented on subsys/bluetooth/host/settings.c:
```diff
@@ -143,8 +144,7 @@ static int set(int argc, char **argv, void *value_ctx)
 	if (!strcmp(argv[0], "name")) {
-		len = settings_val_read_cb(value_ctx, &bt_dev.name,
-					   sizeof(bt_dev.name) - 1);
+		len = read(store, &bt_dev.name, sizeof(bt_dev.name) - 1);
```
There was no check, so I didn't add it

Laczen commented on subsys/bluetooth/host/settings.c:
```diff
@@ -159,8 +159,7 @@ static int set(int argc, char **argv, void *value_ctx)
 	if (!strcmp(argv[0], "irk")) {
-		len = settings_val_read_cb(value_ctx, bt_dev.irk,
-					   sizeof(bt_dev.irk));
+		len = read(store, bt_dev.irk, sizeof(bt_dev.irk));
```
There was no check, so I didn't add it.

Qbicz commented on samples/subsys/settings/custom/src/main.c:
```diff
@@ -0,0 +1,189 @@
+#endif /* CONFIG_SETTINGS_DEFAULT_NONE */
+
+/* 1000 msec = 1 sec */
+#define SLEEP_TIME      2000
```
K_SECONDS(2)

Qbicz commented on samples/subsys/settings/default/src/main.c:
```diff
@@ -0,0 +1,159 @@
+#endif /* CONFIG_SETTINGS_DEFAULT_FS */
+
+/* 1000 msec = 1 sec */
+#define SLEEP_TIME      2000
```
K_SECONDS(2)

Qbicz commented:
@Laczen any updates on this? Is there anything blocking right now?

nvlsianpu commented:
@Qbicz - I need to review this deeply. Come to my desk - we can talk about your opinion.

Qbicz commented:
@nvlsianpu thank you for information. For me as a user of API changes look reasonable, especially loading the information in `set` routines.

Laczen commented:
@Qbicz, there is no need to hurry this. I am still waiting for remarks from @jhedberg and @nvlsianpu. After this I need to convert the tests that are available for the settings subsystem. I know @fnde-ot had a nice experience with the rewritten module.

carlescufi commented:
@Laczen I know that @nvlsianpu is now going to look at your PR in-depth. In the meantime, can you please rebase it?

nvlsianpu reviewed (changes requested):
more review soon.

nvlsianpu commented on subsys/settings/settings_fcb.c:
```diff
@@ -0,0 +1,619 @@
+		len_name = min(SETTINGS_MAX_NAME_LEN + SETTINGS_EXTRA_LEN + 1,
+				loc.loc.fe_data_len);
+
+		rc = settings_fcb_read(&loc, 0, name, &len_name);
```
name begin address is always aligned while end might not.
I recommend to Introduce intermediate function for read without care about alignment and use it elsewhere.

nvlsianpu commented on subsys/settings/settings_fcb.c:
```diff
@@ -0,0 +1,619 @@
+			       buf, *len);
+}
+
+static ssize_t settings_fcb_read_fn(void *back_end, void *data, size_t len)
```
Oh - I miss that value data inside key-value record  might begin at unaligned offset.
Similar as below I recommend to Introduce intermediate function for read without care about alignment and use it elsewhere.

nvlsianpu commented:
I see that fcb back-end rewrites much of previous functionality of settings_line.c file. Unfortunately it has few issues. I think it would be better too keep settings_line.c file with only susbset of usefull function (settings_line_name_read, settings_line_raw_read, settings_line_raw_read_until...), adapt this functions to current needs und use in fcb back-end implementation file. This also makes the review easier, and will keep thei rutines independent, so easy to reuse once needed.

Laczen commented:
@nvlsianpu, thanks for the review. Regarding the fcb issues I can understand what you are saying but there is one main problem that I cannot change:

a. fcb requires the settings module to request space for writing an item, this should be of correct (real) required size in order to read it back later.
b. if sufficient space is found the settings module can write the data.
c. the amount of data that is written can never be of size larger than the requested because you might break the fcb file system. Unless fcb tells you the amount of data to be written.

So to solve this problem I see two possibilities:
1. fcb returns the length that is supposed to be written and then alignment could be done in the settings module.
2. fcb does the correct alignment itself and does not eat any information from the stack to write to flash.
According to me option 2 should be the correct one, but I can also live with option 1. Using the settings module to try and correct this by just adding some data is a bad idea.

Laczen commented on subsys/settings/settings.c:
```diff
@@ -0,0 +1,108 @@
+}
+
+int settings_register(struct settings_handler *handler)
+{
```
I have seen the commit that changed this, however I don't see a reason why the same name could not be used several times. It can be advantageous to register several handlers with the same name, each of the set routines would be called and could provide handling of different items under this same name.

Laczen replied:
You are correct, it will only call the first settings_handler that is registered with this name. I will add the check.

Laczen commented on subsys/settings/Kconfig:
```diff
@@ -6,88 +6,137 @@
 	  Magic 32-bit word for to identify valid settings area
 
-config SETTINGS_FCB_FLASH_AREA
+config SETTINGS_DEFAULT_FCB_FLASH_AREA
```
yes, will be part of the rebase work.

Laczen commented on subsys/settings/settings_fcb.c:
```diff
@@ -0,0 +1,619 @@
+{
+
+	if (off >= entry_ctx->loc.fe_data_len) {
+		return -EINVAL;
```
should this not just be:
```
if (off > entry_ctx->loc.fe_data_len) {
		return -EINVAL;
```

nvlsianpu replied:
After rethinking -depends form how caller use this function -previously it wast possible to call it for heck whether read-out is at the end of record.

nvlsianpu commented on subsys/settings/settings_nvs.c:
```diff
@@ -0,0 +1,229 @@
+
+	write_name = true;
+	/* Find out if we are doing a delete */
+	delete = false;
```
hmm: maybe remove `delete` and use `value == NULL` as the deletion flag:
```suggestion
	if ( val_len == 0) {
		value == NULL;
	}
```

Laczen replied:
ok, but then I would rather use val_len, then I can do a ```if (!val_len)```. I think for an understanding of the code the ```delete``` is better. It does a translate between ```val_len = 0``` or ```value == NULL``` at the beginning of the routine, so the reader knows what is happening.

nvlsianpu replied:
this is up to you - you can ignore this comment even

nvlsianpu commented on subsys/settings/settings_nvs.c:
```diff
@@ -0,0 +1,229 @@
+	write_name_id = cf->last_name_id + 1;
+
+	while (1) {
+
```
rm empty line

Laczen replied:
ok

nvlsianpu commented on subsys/settings/settings_nvs.c:
```diff
@@ -0,0 +1,229 @@
+
+		name_id--;
+		if (name_id == NVS_NAMECNT_ID) {
+			break;
```
doesn't this mean that nvs-back-end has no more spare name ID?

Laczen replied:
No, when reading items the name-id's are "traveled" in reverse order, when reaching ```NVS_NAME_CNT_ID``` there are no more name-id's to read.

nvlsianpu commented on subsys/settings/settings_nvs.c:
```diff
@@ -0,0 +1,229 @@
+
+
+		if ((delete) && (name_id == cf->last_name_id)) {
+			cf->last_name_id--;
```
on,y place where ID are freed. We can come to situation when some lower ID are free as well, but will be not available for usage. This mean that it is possible to create runtime situation where the backend will keep only one key-value pair and will be unable to store more than that. I think need to add algorithm for finding this lower spare IDs.

Laczen replied:
Yes this is the only place where ID's are freed. However we cannot get into a situation where there are no more id's (except when all 16383 id's are in use) because when saving new items it will go through the name-ids in forward order, whenever a unused name-id is detected this will be used to store the name.

nvlsianpu commented on subsys/settings/settings_nvs.c:
```diff
@@ -0,0 +1,229 @@
+	}
+
+	/* write the value */
+	rc = nvs_write(&cf->cf_nvs, write_name_id + NVS_NAME_ID_OFFSET,
```
I think it is possible to overwrite the last key if backend is at ID end.

Laczen replied:
Yes this is true, but it does this when you have reached 16383 entries.

nvlsianpu reviewed (changes requested):
i sppot

Laczen commented on subsys/settings/settings_file.c:
```diff
@@ -0,0 +1,613 @@
+/*
+ * Called to advance to a next entry in a file
+ *
+ * Items are stored as: string - binary where
```
Because it is much easier to construct a read function if it is stored as ```name=<len><value>```, and if encoding would be used (which makes sense for file backend) the storage would be ```name=<encoded value>```. The ```<len><value>``` construct is just a method of encoding.

nvlsianpu replied:
If it is more convince for you - that's ok.

Laczen commented on subsys/settings/settings_file.c:
```diff
@@ -0,0 +1,613 @@
+	entry->off_value = entry->off_name;
+	entry->len_value = 0;
+
+	rc = fs_seek(file, 0, FS_SEEK_END);
```
Yes it does, it will tell that you are trying to get data past the end of file, which you cannot. It returns an -ENOENT error which allows to determine that you have reached the last entry.

nvlsianpu replied:
I see. But doesn't  actually sf_seek returns an error in this case as well.

Laczen commented on subsys/settings/settings_file.c:
```diff
@@ -0,0 +1,613 @@
+	if (rc) {
+		return rc;
+	}
+	if (fs_tell(file) <= entry->off_name) {
```
See above.

nvlsianpu commented:
> The user can setup any combination of backends when not using a
default backend (for this a settings_backend_init() routine has to be
provided).

So it is possible to use a few destination back-end in the same time? Or just read from a few?

Laczen commented:
> > The user can setup any combination of backends when not using a
> > default backend (for this a settings_backend_init() routine has to be
> > provided).
> 
> So it is possible to use a few destination back-end in the same time? Or just read from a few?

There is only one destination back-end, and multiple sources. However you can change the destination backend, this allows you to write to a different backend.

Laczen commented on subsys/settings/settings_fcb.c:
```diff
@@ -0,0 +1,619 @@
+			continue;
+		}
+
+		len_name = strchr(name, '=') - name;
```
ok

Laczen commented on subsys/settings/settings_default.c:
```diff
@@ -0,0 +1,108 @@
+#elif defined(CONFIG_SETTINGS_DEFAULT_NVS)
+#include <device.h>
+static struct settings_nvs default_settings = {
+	.cf_nvs.sector_size = FLASH_ERASE_BLOCK_SIZE *
```
ok

Laczen commented on subsys/settings/Kconfig:
```diff
@@ -6,88 +6,137 @@
-	string "Serialization directory"
-	default "/settings"
-	depends on SETTINGS && SETTINGS_FS
+config SETTINGS_DEFAULT_FS_MNT
```
No, this should be the mount point, when trying to figure out the old settings with nffs I had a lot of difficulty finding out where the mount point should be. By defining it like this it is possible to define nffs with the mount point and then later use this mount point for your file.

nvlsianpu replied:
I typical situation settings will coexisting with other modules in the FS. Otherwise doesn't make sense to us FS backend at all (because NVS and FCB is much smaller and faster tha NFFS) . FS should be mounted before settings initialization. So this should be directory.

Laczen replied:
Please have a good look at this, I don't think the current definitions make any sense. The problem is that the filename and filedir should contain the mount point. I have no idea how to solve this, except by specifying the mount point in the config file. Maybe you have a better idea, but according to me the current kconfig does not work.

nvlsianpu replied:
> The problem is that the filename and filedir should contain the mount point.

Yes - and it should be like that. Settings need absolute path to the file. Directory is place when we can create temporary file while compressing the settings file. I just check that old file name and directory name didn't contain mount points - it was jut because multi-fs was introduced after FS back-end. (i think this might have mislead you)

Laczen replied:
OK, the reworked settings does not need a directory name, so there is no need to define it.
Regarding the mount point: when the default backend is selected to be of type file everything should work regardless of how someone wishes to use it. When the correct mount point for the app is selected the settings ```CONFIG_SETTINGS_DEFAULT_FS_MNT``` and the ```CONFIG_SETTINGS_DEFAULT_FILENAME``` can be changed to reflect the chosen mount point.
I am open to a proposal that changes this, but it should:
a. Work out of the box for a sample,
b. Be easily adapted to user requirements.

nvlsianpu replied:
I didn't catch that this is for sample only, and FS backend needs only SETTINGS_DEFAULT_FS_FILE.
Pleas remove this kconfig form here - as this is not for the subsys but for samles.

Laczen replied:
This is not really for sample only, maybe I should redefine the SETTINGS_DEFAULT_FS_FILE as the concat of SETTINGS_DEFAULT_FS_MNT and SETTINGS_DEFAULT_FS_FILE (without the mount point). Kconfig can then be used to set SETTINGS_DEFAULT_FS_MNT according to the user needs.

nvlsianpu replied:
Or keep only SETTINGS_DEFAULT_FS_FILE with absolute path. In sames extract the mount point from it. Advantage is less properties to see for the end-user. I thin it is also intuitive.

nvlsianpu replied:
Ok - lets do this as you proposed (SETTINGS_DEFAULT_FS_MNT and SETTINGS_DEFAULT_FS_FILE )

Laczen commented:
@fnde-ot, @nvlsianpu, @jhedberg, while doing the rework after review on the proposed settings rework I have come up with something that you might find a good idea. At the moment when processing a item that is read from storage the settings module splits up the entire name string into separate arguments. Suppose we don't let it work like this but only to process up to the next "/" separation, call the correct handler and pass on the string to the handler without the just processed item. The settings handler could
then process the remaining string and call its own "sub-handlers". Maybe a bit more elaborated (fictive example):

variable "/bt/btmesh/rpl/0x0000=0":   
settings_handler "bt_set" is called with the arguments("btmesh","rpl/0x0000", ...)
this call a routine "btmesh_set" with the arguments("rpl","0x0000", ...)

The routine to split up the name string into the next variable and the remaining string would be provided by settings.c and reused by all modules that need it.

Maybe I am overlooking something so I would like your opinion on this.

fnde-ot commented:
> Maybe I am overlooking something so I would like your opinion on this.

I don't see anything against, it would slightly reduce stack usage (single char* vs int argc and char *argv[8]) and allow for a bit more flexibility at the handler level on how to interpret the rest of the key.

You should try and see how the module handlers look after the change.

nvlsianpu commented:
> Maybe I am overlooking something so I would like your opinion on this.
...
      You should try and see how the module handlers look after the change.

I think this PR scope is big enough - this can be eventually done in subsequent PR.
see general no big obstacles, but then the handler needs to interpret sub-key on its own, so this will complicate code a little. Amount of stack consumed on args now is not so big. I also think this might collide with requested subtree operations https://github.com/zephyrproject-rtos/zephyr/issues/12969

Laczen commented:
> > Maybe I am overlooking something so I would like your opinion on this.
> > ...
> > You should try and see how the module handlers look after the change.
> 
> I think this PR scope is big enough - this can be eventually done in subsequent PR.
> see general no big obstacles, but then the handler needs to interpret sub-key on its own, so this will complicate code a little. Amount of stack consumed on args now is not so big. I also think this might collide with requested subtree operations #12969
> 
> so I the end I see obstacles.

OK, I will leave it for a future PR,

@fnde-ot: as this would change the use of SETTINGS_MAX_NAME_LEN, ... I will leave it as it is. 

@nvlsianpu, #12969 should be solved by the present PR as each handler can be disabled, so it is possible to delay loading of dependent handlers. A small routine could be added to store the current handlers, disable all but the requested, call load, and restore the previously enabled handlers. So there would be no collision.

nvlsianpu commented:
> #12969 should be solved by the present PR as each handler can be disabled, so it is possible to delay loading of dependent handlers. A small routine could be added to store the current handlers, disable all but the requested, call load, and restore the previously enabled handlers. So there would be no collision.


I believe something more have to be implemented as we need to filter exact keys to any depth.

Laczen commented:
> > #12969 should be solved by the present PR as each handler can be disabled, so it is possible to delay loading of dependent handlers. A small routine could be added to store the current handlers, disable all but the requested, call load, and restore the previously enabled handlers. So there would be no collision.
> 
> I believe something more have to be implemented as we need to filter exact keys to any depth.

I don't know, I think for module dependency the solution with proposed disable is perfect. On the other side I don't see any problem in adding a name to settings_load() to allow the calling of a settings_load of a subtree, it is just a matching of the read name and requested name and then to call the appropriate handler, this could even be done down to the variable name. However... every call to settings_load(name) will need to go through all stored items and this might be slow. At the moment I don't really see a use case for settings_load(name), there is one exception: I see a use case for reading variables directly, but this would bypass the settings module completely.
Anyhow, neither of these things is limited by the keeping of the name string while stepping down in the settings_handler.

nvlsianpu commented:
> At the moment I don't really see a use case for settings_load(name), there is one exception: I see a use case for reading variables directly, but this would bypass the settings module completely.

We already have such needs - will introduce them as next steps.
Handler disabling feature you added will be fruitful with them as well.
After rethinking se no collision as well.

Laczen committed:
subsys/settings: Modifications applied to sample

The sample: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app
has been updated to the reworked settings module.

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

Laczen committed:
subsys/settings: Modifications to prj.conf files

Several samples that are using the settings module needed
an update: the configuration option CONFIG_SETTINGS_FCB now means
that support for FCB is included. To setup a default FCB backend
the configuration option is now CONFIG_SETTINGS_DEFAULT_FCB. The change
has been applied he sample to the samples/tests that use this.

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

Laczen commented:
@nvlsianpu, @fnde-ot, @Qbicz with the last update I think all review remarks have been taken into account.
@carlescufi, also the rebase was done.

nvlsianpu commented:
@Laczen I will revisit tomorrow. 
Just want to mention that the subsystem tests need pass (I don't se any alignment here).

nvlsianpu commented on samples/subsys/settings/default/README.rst:
```diff
@@ -0,0 +1,31 @@
+Building and Running
+********************
+
+This sample can be found under :file:`samples/subsys/settings/default` in the Zephyr tree.
```
^^ still not fixed

Laczen commented on subsys/settings/Kconfig:
```diff
@@ -21,71 +18,125 @@ module-str = settings
-	depends on SETTINGS && SETTINGS_FS
+	depends on SETTINGS && SETTINGS_DEFAULT_FS
 	help
 	  Full path to the default settings file.
```
Agreed, will change this to: ``Path of the default settings file``

Laczen commented on samples/subsys/settings/custom/src/main.c:
```diff
@@ -0,0 +1,197 @@
+	int rc;
+	struct flash_pages_info info;
+
+	rc = flash_get_page_info_by_offs(device_get_binding(DT_FLASH_DEV_NAME),
```
Correct, this should be DT_FLASH_AREA_STORAGE_DEV once PR #13954 has been merged.

Laczen commented on subsys/settings/settings_default.c:
```diff
@@ -0,0 +1,120 @@
+	int rc;
+	struct flash_pages_info info;
+
+	rc = flash_get_page_info_by_offs(device_get_binding(DT_FLASH_DEV_NAME),
```
Correct, this should be DT_FLASH_AREA_STORAGE_DEV once PR #13954 has been merged.

nvlsianpu replied:
You can use now recently meregd: [API for get driver structure belongs to the flash_area](https://github.com/zephyrproject-rtos/zephyr/pull/13303)

Laczen replied:
> You can use now recently meregd: [API for get driver structure belongs to the flash_area](https://github.com/zephyrproject-rtos/zephyr/pull/13303)
Yes, but this requires flash_map, and I don't have any spare ram to accommodate the flash_map. PR#13954 makes it clean and simple :).

nvlsianpu replied:
flash_map lands in ROM.
One of Objective for flash_map is central source of information - so it can be even modified/shared in product lifetime..

nvlsianpu commented on subsys/settings/settings_fcb.c:
```diff
@@ -0,0 +1,642 @@
+	/* If there is more data read it */
+	if (len) {
+		tmp_len = len;
+		rc = settings_fcb_read(rd_fn_arg->entry_ctx, off, data8,
```
Data can have end-up at unaligned position. 
Fcb data field always begin at aligned address (how ever this function can be used for readout from the middle of it, which is processed properly a few lines above), It also physically ends at aligned bloc end, but some very last bytes might be just the filler.
```
aligned ----------------------
 |                            |
\/                            \/
 | payload | filler <0,wbs-1> |
      ^    ^
      |    |
      |    payload data end might be unaligned
      |    
      - redout might be requested from unaligned location
        as well
```

Laczen replied:
This is exactly what the routine does, it starts the read from the previous alignment, reads up to the next, if all data is inside the alignment the routine stops. if there still is data it reads until the end of the data. The routine never reads out of boundary of the fcb entry and returns the correct amount of data. Whenever a read is done outside the fcb entry an error is generated. The routine also never reads outside the payload, it shouldn't because there is no data there.

nvlsianpu replied:
- user might call this function with any length he want.
- fcb encodes data field len as len of payload, not payload + filler len. So that len is aligned in this read is not guaranteed.

Laczen replied:
> * user might call this function with any length he want.
Yes, but the routine will only read/return what is inside the payload. The length that is read is returned.  
> * fcb encodes data field len as len of payload, not payload + filler len. So that len is aligned in this read is not guaranteed.
Why should len be aligned ? You can read whatever amount of bytes starting from an aligned position.

nvlsianpu replied:
>Why should len be aligned ? You can read whatever amount of bytes starting from an aligned position.
a flash driver  might serve only aligned access, both on write as on read. And this is true for many drivers in zephyr.

nvlsianpu replied:
apparently settings_fcb_read need correction as well.

nvlsianpu replied:
^^ looks like length read be aligned to wbs - and this is probably a bug in zephyr mainline as well :(

Laczen replied:
I had a look at the drivers, most of them even allow unaligned reads, I don't know the drivers you are talking about. No need to make any changes to zephyr mainline, maybe a driver needs change.

nvlsianpu replied:
SRM drivers for example

Laczen replied:
> SRM drivers for example

I can't find any SRM drivers in the flash directory. Maybe you mean stm32, they do not put any alignment requirements on flash reads.

nvlsianpu replied:
right - was STM32, indeed it have must been upgraded since I had looked at them before.
QMSI has requirement for 4 B alignment.
MCUX has not.
Geco has not.
SAM has not.
NIOS2 has requirement for 4 B alignment.
NRF has not.

Laczen replied:
> right - was STM32, indeed it have must been upgraded since I had looked at them before.
> QMSI has requirement for 4 B alignment.
> MCUX has not.
> Geco has not.
> SAM has not.
> NIOS2 has requirement for 4 B alignment.

NIOS has no length alignment, only start

> NRF has not.

So the only one left is QMSI, that could easily be changed to remove the length alignment requirement.

nvlsianpu replied:
I didn't check SPI drivers -  if w25qxxdv has any requirement, it mus be depend on spec. (I didn't catch anything in its code).
And there is also spi_nor driver.

Laczen replied:
> I didn't check SPI drivers - if w25qxxdv has any requirement, it mus be depend on spec. (I didn't catch anything in its code).

No alignment requirements

> And there is also spi_nor driver.

No alignment requirements

Laczen commented:
After careful thinking about this PR I have decided to remove it. I would like to thank @dbkinder and @fnde-ot for their time and effort to improve the proposed module. As people are trying to force me into enabling flash_map in applications that use settings with the NVS backend I cannot stand behind this.

nvlsianpu commented:
That's bad.
Oh well. Can leave without this
You could left PR open and not to delete the branch - I or someone else could finish up it. Of course this was up to you.
